src/evalfunc.c

```diff
@@ -5452,6 +5455,181 @@ f_getpos(typval_T *argvars, typval_T *rettv)
+	*(p + bd->endspaces) = NUL;
+    }
+    return ret;
+}
+
+/*
+ * "getregion()" function
+ */
+    static void
+f_getregion(typval_T *argvars, typval_T *rettv)
+{
+    linenr_T		lnum;
+    oparg_T		oap;
+    struct block_def	bd;
+    char_u		*akt;
+    int			inclusive = TRUE;
+    int			fnum = -1;
+    pos_T		p1, p2;
+    pos_T		*fp = NULL;
+    char_u		*str1, *str2;
+    int			is_visual = FALSE;
+
+    if (rettv_list_alloc(rettv) == FAIL)
+	return;
+
+    virtual_op = virtual_active();
+
+    if (check_for_string_arg(argvars, 0) == FAIL
+	    || check_for_string_arg(argvars, 0) == FAIL)
+	return;
+
+    // NOTE: var2fpos() returns static pointer.
+    fp = var2fpos(&argvars[0], TRUE, &fnum, FALSE);
+    if (fp == NULL)
+	return;
+    p1 = *fp;
+
+    fp = var2fpos(&argvars[1], TRUE, &fnum, FALSE);
+    if (fp == NULL)
+	return;
+    p2 = *fp;
+
+    str1 = tv_get_string(&argvars[0]);
+    str2 = tv_get_string(&argvars[1]);
+
+    if (!LT_POS(p1, p2))
+    {
+	// swap position
+	pos_T p;
+
+	p = p1;
+	p1 = p2;
+	p2 = p;
+    }
+
+    is_visual = (str1[0] == 'v' && str1[1] == NUL)
+	|| (str2[0] == 'v' && str2[1] == NUL);
+    if (is_visual)
+    {
+	if (VIsual_mode == 'v')
+	{
+	    // handle 'selection' == "exclusive"
+	    if (*p_sel == 'e' && !EQUAL_POS(p1, p2))
+	    {
+		if (p2.coladd > 0)
+		    p2.coladd--;
+		else if (p2.col > 0)
+		{
+		    p2.col--;
+		    mb_adjustpos(curbuf, &p2);
+		}
+		else if (p2.lnum > 1)
+		{
+		    p2.lnum--;
+		    p2.col = (colnr_T)STRLEN(ml_get(p2.lnum));
+		    if (p2.col > 0)
+		    {
+			p2.col--;
+			mb_adjustpos(curbuf, &p2);
+		    }
+		}
+	    }
+	    // if fp2 is on NUL (empty line) inclusive becomes false
+	    if (*ml_get_pos(&p2) == NUL && !virtual_op)
+	    {
+		inclusive = FALSE;
+	    }
+	}
+	else if (VIsual_mode == Ctrl_V)
+	{
+	    colnr_T sc1, ec1, sc2, ec2;
+
+	    getvvcol(curwin, &p1, &sc1, NULL, &ec1);
+	    getvvcol(curwin, &p2, &sc2, NULL, &ec2);
+	    oap.motion_type = OP_NOP;
+	    oap.inclusive = TRUE;
+	    oap.start = p1;
+	    oap.end = p2;
+	    oap.start_vcol = MIN(sc1, sc2);
+	    if (*p_sel == 'e' && ec1 < sc2 && 0 < sc2 && ec2 > ec1)
+		oap.end_vcol = sc2 - 1;
+	    else
+		oap.end_vcol = MAX(ec1, ec2);
+	}
+    }
+
+    // Include the trailing byte of a multi-byte char.
+    const int l = utfc_ptr2len((char_u *)ml_get_pos(&p2));
+    if (l > 1)
+	p2.col += l - 1;
+
+    for (lnum = p1.lnum; lnum <= p2.lnum; lnum++)
+    {
+	if (is_visual && VIsual_mode == 'V')
+		akt = vim_strsave(ml_get(lnum));
+	else if (is_visual && VIsual_mode == Ctrl_V)
+	{
+		block_prep(&oap, &bd, lnum, FALSE);
+		akt = block_def2str(&bd);
+	}
+	else if (p1.lnum < lnum && lnum < p2.lnum)
+	    akt = vim_strsave(ml_get(lnum));
+	else
+	{
+	    charwise_block_prep(p1, p2, &bd, lnum, inclusive);
+	    akt = block_def2str(&bd);
+	}
+
+	if (akt == NULL || list_append_string_move(rettv->vval.v_list, akt) == FAIL)
+	{
+	    list_free(rettv->vval.v_list);
+	    break;
+	}
+    }
+    virtual_op = MAYBE;
```

Comment:
Instead of overwriting `virtual_op`, should the old value be saved at the beginning of this function and restored here?